csrc/id_model/contiguity.cpp

```diff
@@ -0,0 +1,146 @@
+
+ContigIDGroups::ContigIDGroups(
+    const std::vector<IterDomain*>& alloc_domains,
+    std::vector<bool> contiguity,
+    const ExprPath<ExprGroup>& path_from_alloc,
+    const ValGraph& graph)
+    : graph_(graph),
+      alloc_domains_(alloc_domains),
+      alloc_contiguity_(std::move(contiguity)),
+      consistent_transform_info_(
+          std::make_unique<const OrderedIdGroupInformation>(
+              OrderedIdGroupInformation::get(
+                  alloc_domains,
+                  path_from_alloc,
+                  graph))) {
+  if (alloc_domains_.empty()) {
+    return;
+  }
+
+  NVF_ERROR(
+      alloc_domains_.size() == alloc_contiguity_.size(),
+      "Arguments don't match ",
+      alloc_domains_.size(),
+      " != ",
+      alloc_contiguity_.size());
+
+  for (const auto index_domain_i : c10::irange(alloc_domains_.size())) {
+    IterDomain* index_domain = alloc_domains_.at(index_domain_i);
+    NVF_ERROR(
+        !index_domain->isBroadcast(),
+        "Broadcast domain should not be an index domain: ",
+        index_domain->toString());
+
+    alloc_to_contig_ids_[index_domain] = graph_.toGroup(index_domain);
+
+    auto alloc_contiguity = alloc_contiguity_.at(index_domain_i);
+
+    if (alloc_contiguity &&
+        index_domain->getIterType() != IterType::GatherScatter) {
+      contig_ids_.emplace(graph_.toGroup(index_domain));
+    }
+  }
+
+  for (const auto& [eg, direction] : path_from_alloc) {
+    // Propagate resize and non-divisible dependencies
+    const auto inputs = direction == Direction::Forward
+        ? graph_.inputGroups(eg)
+        : graph_.outputGroups(eg);
+    const auto outputs = direction == Direction::Forward
+        ? graph_.outputGroups(eg)
+        : graph_.inputGroups(eg);
+    if (std::any_of(inputs.begin(), inputs.end(), [&](const ValGroup& inp) {
+          return resize_deps_.count(inp) > 0;
+        })) {
+      for (const auto& out : outputs) {
+        resize_deps_.insert(out);
+      }
+    }
+
+    dispatch(eg, direction);
+  }
+}
+
+void ContigIDGroups::handle(Merge* merge, Direction direction) {
+  // Only forward direction is supported for now
+  if (direction != Direction::Forward) {
+    return;
+  }
+
+  // If output is not consistently ordered or doesn't solely consume all
+  // allocation domains in its dependencies, then it can't be a contiguously
+  // indexable iterdomain.
+  if (!consistent_transform_info_->isConsistentlyOrdered(merge->out())) {
+    return;
+  }
+
+  if (!consistent_transform_info_->exclusivelyConsumesAllocs(merge->out())) {
+    return;
+  }
+
+  // Check allocation domains for contiguity
+  auto alloc_ids_it = consistent_transform_info_->findAllocIDs(merge->out());
+  VectorOfUniqueEntries<IterDomain*> alloc_ids = alloc_ids_it->second;
+  for (auto alloc_id_i : c10::irange(alloc_domains_.size())) {
+    auto alloc_id = alloc_domains_[alloc_id_i];
+    if (!alloc_ids.has(alloc_id)) {
+      continue;
+    }
+    alloc_ids.erase(alloc_id);
```

Comment:
Tiny comment: we could do this in a single check if we changed `VectorOfUniqueEntries::erase` to return the number of removed elements: https://github.com/NVIDIA/Fuser/blob/e275d8fa337cdc19b7488b93906dcf13e9b52172/csrc/disjoint_set.h#L203-L212 by just returning the result at line 211 there then modifying this code to
```suggestion
    if (alloc_ids.erase(alloc_id) == 0) {
      continue;
    }
```